chore: Update README.md

kramaranya opened the pull request:
I've updated README with initial content. We shall expand this further before 0.1 release - just wanted to get something more substantial in place for now!

Also borrowed the idea about contributors from Feast and Llama Stack, I think it's really cool :)

/assign @kubeflow/wg-training-leads  @andreyvelich @franciscojavierarceo @Electronic-Waste @astefanutti @briangallagher @szaher

Electronic-Waste commented on README.md:
```diff
@@ -1,10 +1,78 @@
+ML applications rather than managing complex infrastrutcure.
 
-Design doc: https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit
+> **Note**: The Kubeflow SDK is in active development. Currently, the SDK provides full support for **Kubeflow
+Trainer** with other components planned for future releases.
```
Can we replace it with the Github style alerts? Maybe it would be better, just FYI.

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

Electronic-Waste reviewed (commented):
@kramaranya Thanks for this! It's amazing. One initial thought for you:)

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+| Component                   | Status | Description                                                    |
+|-----------------------------|--------|----------------------------------------------------------------|
+| **Kubeflow Trainer**        | ✅ **Available** | Train and fine-tune models with various frameworks   |
+| **Kubeflow Katib**          | 🚧 Planned | Hyperparameter optimization                                |
+| **Kubeflow Pipelines**      | 🚧 Planned | Build, run, and track ML workflows                         |
+| **Kubeflow Spark Operator** | 🚧 Planned | Run Apache Spark jobs on Kubernetes                        |
+| **Kubeflow Model Registry** | 🚧 Planned | Manage model artifacts, versions and ML artifacts metadata |
+| **Kubeflow Feast**          | 🚧 Planned | Feature store for ML feature management and serving        |
+| **Kubeflow KServe**         | 🚧 Planned | Model deployment and serving infrastructure                |
```
Can we keep only Katib, Pipelines, and Registry here for now ?
We should have discussions on long-term approach for supported projects.

andreyvelich commented on README.md:
````diff
@@ -1,10 +1,78 @@
+pip install git+https://github.com/kubeflow/sdk.git@main
+```
+
+## Supported Components
````
```suggestion
## Supported Kubeflow Projects
```

kramaranya replied:
Updated

andreyvelich commented on README.md:
````diff
@@ -1,10 +1,78 @@
+```bash
+pip install git+https://github.com/kubeflow/sdk.git@main
+```
+
````
Maybe it worths to add simple getting started example, like in LangGraph: https://pypi.org/project/langgraph/

```py
from kubeflow.trainer import TrainerClient, CustomTrainer

def get_torch_dist():
    import os
    import torch
    import torch.distributed as dist

    dist.init_process_group(backend="gloo")
    print(f"PyTorch Distributed Environment")
    print(f"WORLD_SIZE: {dist.get_world_size()}")
    print(f"RANK: {dist.get_rank()}")
    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")

# Create the TrainJob.
job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
    trainer=CustomTrainer(
        func=get_torch_dist,
        num_nodes=3,
        resources_per_node={
            "cpu": 2,
        },
    ),
)

# Wait for TrainJob to complete.
TrainerClient().wait_for_job_status(job_id)

# Print TrainJob logs.
print(TrainerClient().get_job_logs(name=job_id, node_rank=0)["node-0"])
```

WDYT @kramaranya ?

kramaranya replied:
Yeah, I've been thinking about this too! I will add that

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+
+- **Slack**: Join our [#kubeflow-trainer](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels) Slack channel
+  <!-- TODO(kramaranya): update to #ml-experience once we kick off -->
+- **Meetings**: Attend the [AutoML and Training Working Group](https://bit.ly/2PWVCkV) bi-weekly meetings
+  <!-- TODO(kramaranya): update with kubeflow sdk meetings once we kick off -->
```
We can change it to Kubeflow SDK call: https://docs.google.com/document/d/1jH2WAX2ePxOfI4JuiVK9nPlesDMiyg67xzLwhpR7wTQ/edit?tab=t.0

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+### Contributing
+
+Kubeflow SDK is a community project and is still under active development. We welcome contributions! Please see our
+[CONTRIBUTING Guide](CONTRIBUTING.md) for details.
```
Since this README is going to publish to PyPI, we might need to use complete URL for the docs.
```suggestion
[CONTRIBUTING Guide](https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md) for details.
```

kramaranya replied:
Good point, updated!

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+- **Simplified ML Workflows**: Abstract away Kubernetes complexity, allowing ML practitioners to work in familiar Python environments
+- **Seamless Integration**: Designed to work together with all Kubeflow projects for end-to-end ML pipelines
+- **Local Development**: First-class support for local development requiring only `pip` installation
+
```
Do we see any value to include persona diagram here ?
https://www.kubeflow.org/docs/components/trainer/overview/#who-is-this-for

kramaranya replied:
Since this diagram is specific to Trainer, I think that makes sense to include it for the first release and then create a more generic version for future releases as we add more KF projects? WDYT @andreyvelich @astefanutti @Electronic-Waste ?

astefanutti replied:
@kramaranya that makes sense to me.

kramaranya replied:
I've added the diagram, let me know if there's anything I should change

andreyvelich replied:
Alternatively, we can re-use diagram from the SDK proposal: https://github.com/kubeflow/community/blob/master/proposals/819-kubeflow-sdk/kubeflow-sdk.drawio.svg

We might need to have better diagram in the future where we say that Kubeflow SDK is the primary interface for AI Practitioners to interact with Kubeflow projects APIs.

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+
+### Kubeflow SDK Benefits
+
+- **Unified Experience**: Single SDK to interact with multiple Kubeflow components through consistent Python APIs
```
```suggestion
- **Unified Experience**: Single SDK to interact with multiple Kubeflow projects through consistent Python APIs
```

kramaranya replied:
Updated! Shall we change https://www.kubeflow.org/docs/components/ to https://www.kubeflow.org/docs/projects/ btw? 🤔

andreyvelich replied:
We had discussion before with @thesuperzapper: https://github.com/kubeflow/website/issues/3982
It might be tricky to maintain redirects for each page if we change URL from `/components` to `/projects`.

kramaranya replied:
I see, thank you

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+### Kubeflow SDK Benefits
+
+- **Unified Experience**: Single SDK to interact with multiple Kubeflow components through consistent Python APIs
+- **Simplified ML Workflows**: Abstract away Kubernetes complexity, allowing ML practitioners to work in familiar Python environments
```
I wouldn't overuse ML workflows term. 

```suggestion
- **Simplified AI Workloads**: Abstract away Kubernetes complexity, allowing AI practitioners to work in familiar Python environments
```

kramaranya replied:
updated

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,78 @@
+<a href="https://github.com/kubeflow/sdk/graphs/contributors">
+  <img src="https://contrib.rocks/image?repo=kubeflow/sdk" />
+</a>
+
+## License
+
+Licensed under the [Apache License 2.0](LICENSE).
```
I don't think that we need it since Licence file presents in root.

kramaranya replied:
Makes sense, updated

andreyvelich reviewed (commented):
Thank you for this @kramaranya!

astefanutti commented:
That'll LGTM after including @andreyvelich feedback.

Thanks @kramaranya!

kramaranya commented on README.md:
```diff
@@ -1,10 +1,78 @@
+<!-- TODO(kramaranya): update when release [![PyPI version](https://img.shields.io/pypi/v/kubeflow?color=%2334D058&label=pypi%20package)](https://pypi.org/project/kubeflow/) -->
+[![Join Slack](https://img.shields.io/badge/Join_Slack-blue?logo=slack)](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels)
 [![Coverage Status](https://coveralls.io/repos/github/kubeflow/sdk/badge.svg?branch=main)](https://coveralls.io/github/kubeflow/sdk?branch=main)
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue)](https://github.com/kubeflow/sdk/blob/main/LICENSE)
```
Thank you! I'll add it
However, why shouldn't we keep the License badge though if we'll publish it to PyPi?

andreyvelich replied:
I just feel that it is redundant since PyPI already renders licence in the Meta section: https://pypi.org/project/kubeflow-trainer-api/#description
Same as GitHub.

andreyvelich replied:
@kramaranya Did you push the commit to remove licence badge ?

kramaranya commented on README.md:
```diff
@@ -1,10 +1,78 @@
 # Kubeflow SDK
 
+<!-- TODO(kramaranya): update when release [![PyPI version](https://img.shields.io/pypi/v/kubeflow?color=%2334D058&label=pypi%20package)](https://pypi.org/project/kubeflow/) -->
```
Yep, that makes sense!

andreyvelich commented on README.md:
````diff
@@ -1,10 +1,78 @@
+Install the Kubeflow SDK using pip:
+
+```bash
+pip install git+https://github.com/kubeflow/sdk.git@main
````
@kramaranya Did you push that change in this PR ?

andreyvelich commented on README.md:
````diff
@@ -1,10 +1,112 @@
+```
+<!-- TODO(kramaranya): update before release pip install -U kubeflow -->
+
+### Run your first distributed training job
````
```suggestion
### Run your first PyTorch distributed job
```

andreyvelich commented on README.md:
```diff
@@ -1,10 +1,112 @@
+
+We couldn't have done it without these incredible people:
+
+<a href="https://github.com/kubeflow/sdk/graphs/contributors">
+  <img src="https://contrib.rocks/image?repo=kubeflow/sdk" />
+</a>
```
@kramaranya Do you know if PyPI is going to render it as well ?

kramaranya replied:
Yep! It works for https://pypi.org/project/llama-stack/ and https://pypi.org/project/feast/

kramaranya commented:
@andreyvelich @astefanutti @Electronic-Waste thank you for the review!
I should have addressed all your comments, let me know if there's anything else should be changed

astefanutti commented:
/lgtm

Thanks @kramaranya!

kramaranya commented:
/rerun

andreyvelich commented:
/retest

kramaranya commented:
@andreyvelich i've rebased it, should be passing now
